formula-lint/src/main/java/com/instacart/formula/lint/WrongFormulaUsageDetector.kt

```diff
@@ -173,4 +210,76 @@ class WrongFormulaUsageDetector : Detector(), Detector.UastScanner {
             typeClass?.qualifiedName == "com.instacart.formula.TransitionContext"
         }
     }
+
+    private fun isKeylessFormulaCallback(
+        context: JavaContext,
+        node: UCallExpression,
+    ): Boolean {
+        val methodName = node.methodName
+        val argsCount = node.valueArgumentCount
+        val methodReceiverType = node.receiverType
+        val typeClass = context.evaluator.getTypeClass(methodReceiverType)
+        val isReceiverFormulaContext = typeClass?.qualifiedName == "com.instacart.formula.FormulaContext"
+        val isCallbackFunction = methodName == "callback"
+        val isKeylessCall = argsCount < 2
+        return isCallbackFunction && isReceiverFormulaContext && isKeylessCall
+    }
+
+    private fun isFormulaContextKeyCall(
+        context: JavaContext,
+        node: UCallExpression,
+    ): Boolean {
+        val methodName = node.methodName
+        val methodReceiverType = node.receiverType
+        val typeClass = context.evaluator.getTypeClass(methodReceiverType)
+        val isReceiverFormulaContext = typeClass?.qualifiedName == "com.instacart.formula.FormulaContext"
+        val isKeyFunction = methodName == "key"
+        return isKeyFunction && isReceiverFormulaContext
+    }
+
+    private fun isWithinKeylessLoop(
+        context: JavaContext,
+        node: UExpression,
+    ): Boolean {
+        var parent = node.uastParent
+        while (parent != null) {
+            if (parent is ULoopExpression) {
+                return true
+            }
```

Comment:
This covers 3 possible types of loop expressions:
- `for`
- `while`
- `do-while`

Reply:
I would add this as a comment in code.